kimchi/src/circuits/polynomials/chacha.rs

```diff
@@ -254,50 +245,105 @@ fn line<F: Field>(nybble_rotation: usize) -> Vec<E<F>> {
 // Gates
 //
 
-/// a += b; d ^= a; d <<<= 16 (=4*4)
-pub fn constraint_chacha0<F: FftField>(alphas: impl Iterator<Item = usize>) -> E<F> {
-    index(GateType::ChaCha0) * E::combine_constraints(alphas, line(4))
+/// Implementation of the ChaCha0 gate
+#[derive(Default)]
+pub struct ChaCha0<F>(PhantomData<F>);
```

Comment:
Oh so you're redefining a `ChaCha0` that is a struct (and the same for each of the `GateTypes`). This explains what I was confused about in our meeting: meaning, I once tried to implement a trait for an enum variant (e.g. GateType::ChaCha0), and rust didn't seem to support that directly. So the reason that you can do this now is ChaCha0 here is a struct. Would it make sense to have some data inside it, instead of just PhantomData? I mean, the corresponding GateType, so that later on when you're defining the `const ARGUMENT_TYPE` you could just access the content of the struct? Well... I guess it is not so much better, it will perhaps take more rows than simply writing manually GateType::ChaCha0 in the very single place that you need it...

Reply:
I'd say it wouldn't make sense to store data that is known at compile-time. The `PhantomData` is a trick to make sure we can use the generic field `F`.

Reply:
Ok thanks :)